### PR TITLE
Fix scheduler hang when pending limit is reached

### PR DESCRIPTION
**Pull request by pollydrag**
## What do these changes do?

Do not add job to `_jobs` list while waiting for a free slot in the pending queue.

## Related issue number

#97, #98 

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] Add a new news fragment into the `CHANGES` folder

### CODE REVIEW

**Comment by Tinche** [comment]
We might be hitting this in production.

**Comment by oleynikandrey** [comment]
Is there any plans to merge this PR?

**Comment by NotoriousRebel** [comment]
@pollydrag Any updates?

**Comment by pollydrag** [comment]
@asvetlov Can you review this, please?

**Comment by AIGeneratedUsername** [comment]
Hi guys, does anybody have a maintained fork of this repository? It may be great to have a maintained fork.

**Comment by AIGeneratedUsername** [comment]
@hlecnt  and other people:

As long as the package is abandoned, I will maintain a fork to help everyone keep using this package without critical bugs. This fix is available:

`pip install aiojobs-fork==1.0.2`

File: aiojobs/_scheduler.py [hunk 1]
```diff
@@ -62,12 +62,16 @@ async def spawn(self, coro):
...
         else:
-            # wait for free slot in queue
-            await self._pending.put(job)
+            try:
+                # wait for free slot in queue
+                await self._pending.put(job)
+            except asyncio.CancelledError:
+                await job.close()
+                raise
+        self._jobs.add(job)
```
Comment (Dreamsorcerer) [inline aiojobs/_scheduler.py:74]: I wonder if it makes more sense to only add jobs when they are started (i.e. "active")?

The only reason I can see for adding pending jobs, is to clean up all the jobs on `.close()`. But, that code also pumps all the jobs out of the queue anyway.
https://github.com/aio-libs/aiojobs/blob/104b4f614b0ea639dc7ebfd590df2faa98ba261d/aiojobs/_scheduler.py#L82-L92

So, maybe we should only track the "active" jobs in this object?

Comment (Dreamsorcerer) [reply aiojobs/_scheduler.py:74]: So, in .close(), I'm thinking you could do something along the lines of:
```
def get_all_pending():
    while not self._pending.empty(): 
         yield self._pending.get_nowait()

all_jobs = self._jobs.union(get_all_pending())
```

Comment (pollydrag) [reply aiojobs/_scheduler.py:74]: Yes, putting only active jobs in _jobs looks interesting. But is it needed here?
This approach requires fixing not only close() but also __iter__/__contains__/etc.
This seems to be more than just a bug fixing and could be done in another task i think

Comment (Dreamsorcerer) [reply aiojobs/_scheduler.py:74]: Ah, right. Maybe it should be kept for iter/contains. Then I think the fix looks reasonable.

File: tests/test_scheduler.py [hunk 2]
```diff
@@ -263,19 +263,54 @@ async def coro(fut):
...
+        await fut
+
+    futures = [asyncio.Future() for _ in range(3)]
+    jobs = []
+
+    async def spawn():
+        for fut in futures:
+            jobs.append(await scheduler.spawn(coro(fut)))
+
+    asyncio.create_task(spawn())
```
Comment (Dreamsorcerer) [inline tests/test_scheduler.py:295]: This task should probably be assigned to a variable and awaited later on (presumably just before the job.wait() calls).

I'm a little surprised the tests don't error due to an unawaited task..

File: tests/test_scheduler.py [hunk 3]
```diff
@@ -263,19 +263,54 @@ async def coro(fut):
...
+    assert scheduler.active_count == 1
+    assert scheduler.pending_count == 1
+
+    for fut in futures:
+        fut.set_result(None)
+
+    for job in jobs:
+        await job.wait()
+
+    assert not len(scheduler)
```
Comment (Dreamsorcerer) [inline tests/test_scheduler.py:308]:
```suggestion
    assert len(scheduler) == 0
```

I think that's more explicit, and also consistent with the other checks below it.

File: tests/test_scheduler.py [hunk 4]
```diff
@@ -263,19 +263,54 @@ async def coro(fut):
...
+
+    futures = [asyncio.Future() for _ in range(3)]
+    jobs = []
+
+    async def spawn():
+        for fut in futures:
+            jobs.append(await scheduler.spawn(coro(fut)))
+
+    asyncio.create_task(spawn())
+    await asyncio.sleep(0.05)
```
Comment (Dreamsorcerer) [inline tests/test_scheduler.py:296]:
```suggestion
    await asyncio.sleep(0)
```
I don't any delay is actually needed. `0` will yield back to the event loop once, then the `.spawn()` calls should only yield to the event loop once the queue is full (if I understand it correctly).

File: tests/test_scheduler.py (hunk 2 above)
Comment (pollydrag) [reply tests/test_scheduler.py]: It's true, thank you. Done

**Review by Dreamsorcerer — APPROVED** [review]
I'm working on type annotations locally. So, not sure if it's easier to merge this before or after those changes.

**Pull request merged**